### PR TITLE
[SPARK-38944][CORE][SQL] Close `UnsafeSorterSpillReader` before `SpillableArrayIterator` throw `ConcurrentModificationException`

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
There will be `UnsafeSorterSpillReader` resource leak(`InputStream` hold by `UnsafeSorterSpillReader` ) when `SpillableArrayIterator` throw `ConcurrentModificationException`, so this pr add resource cleanup process before `UnsafeSorterSpillReader` throws `ConcurrentModificationException`.

### Why are the changes needed?
Fix file resource leak.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
- Pass GA
- Add new check in `ExternalAppendOnlyUnsafeRowArraySuite`

run command:

```
mvn clean install -pl sql/core -am -DskipTests
mvn clean test -pl sql/core -Dtest=none -DwildcardSuites=org.apache.spark.sql.execution.ExternalAppendOnlyUnsafeRowArraySuite
```

**Before**

```
- test iterator invalidation (with spill) *** FAILED ***
  org.apache.spark.io.ReadAheadInputStream@478b0739 did not equal null (ExternalAppendOnlyUnsafeRowArraySuite.scala:397)
Run completed in 9 seconds, 652 milliseconds.
Total number of tests run: 14
Suites: completed 2, aborted 0
Tests: succeeded 13, failed 1, canceled 0, ignored 0, pending 0
*** 1 TEST FAILED ***
```

**After**

```
Run completed in 8 seconds, 535 milliseconds.
Total number of tests run: 14
Suites: completed 2, aborted 0
Tests: succeeded 14, failed 0, canceled 0, ignored 0, pending 0
All tests passed.
```

### CODE REVIEW

**Comment by LuciferYang** [comment]
cc @HyukjinKwon @srowen

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ExternalAppendOnlyUnsafeRowArray.scala [hunk 1]
```diff
@@ -228,6 +234,11 @@ private[sql] class ExternalAppendOnlyUnsafeRowArray(
       currentRow.pointTo(iterator.getBaseObject, iterator.getBaseOffset, iterator.getRecordLength)
       currentRow
     }
+
+    override protected def closeIfNeed(): Unit = iterator match {
```
Comment (srowen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ExternalAppendOnlyUnsafeRowArray.scala:238]: IfNeed -> IfNeeded

**Review by srowen — CHANGES_REQUESTED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/execution/ExternalAppendOnlyUnsafeRowArraySuite.scala [hunk 2]
```diff
@@ -56,7 +59,7 @@ class ExternalAppendOnlyUnsafeRowArraySuite extends SparkFunSuite with LocalSpar
   private def insertRow(array: ExternalAppendOnlyUnsafeRowArray): Long = {
     val valueInserted = random.nextLong()
 
-    val row = new UnsafeRow(1)
+    val row = new UnsafeRow(  1)
```
[the remark this answers is not in the document]
Comment (LuciferYang) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/ExternalAppendOnlyUnsafeRowArraySuite.scala]: done

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ExternalAppendOnlyUnsafeRowArray.scala (hunk 1 above)
Comment (LuciferYang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/ExternalAppendOnlyUnsafeRowArray.scala]: [c8f6f08](https://github.com/apache/spark/pull/36262/commits/c8f6f0871f5875433d5dc951805a7e40d875b7ee) fix this

**Review by srowen — APPROVED** [review]

**Comment by srowen** [comment]
Merged to master

**Comment by LuciferYang** [comment]
thanks @srowen

**Pull request closed without merging**